pkg/pgmodel/ingestor/ingestor.go

```diff
@@ -47,22 +49,46 @@ func NewPgxIngestorForTests(conn pgxconn.PgxConn) (*DBIngestor, error) {
 	return NewPgxIngestor(conn, c, s, &Cfg{})
 }
 
+// Request is a write request for timeseries into Promscale.
+type Request struct {
+	Tenant string
+	Ts     []prompb.TimeSeries
+	Req    *prompb.WriteRequest
+}
+
+// SetWriteAuthorizer sets the write authorizer for multi-tenancy.
+func (ingestor *DBIngestor) SetWriteAuthorizer(wAuth tenancy.WriteAuthorizer) {
+	ingestor.wAuth = wAuth
+}
+
 // Ingest transforms and ingests the timeseries data into Timescale database.
 // input:
 //     tts the []Timeseries to insert
 //     req the WriteRequest backing tts. It will be added to our WriteRequest
 //         pool when it is no longer needed.
-func (ingestor *DBIngestor) Ingest(tts []prompb.TimeSeries, req *prompb.WriteRequest) (uint64, error) {
-	var totalRows uint64
-	dataSamples := make(map[string][]model.Samples)
-	for i := range tts {
-		ts := &tts[i]
+func (ingestor *DBIngestor) Ingest(r Request) (uint64, error) {
+	var (
+		err       error
+		labels    []prompb.Label
+		totalRows uint64
+
+		dataSamples = make(map[string][]model.Samples)
+	)
+	for i := range r.Ts {
+		ts := &r.Ts[i]
 		if len(ts.Samples) == 0 {
 			continue
 		}
+		labels = ts.Labels
+		if ingestor.wAuth != nil {
+			labels, err = ingestor.wAuth.VerifyAndApplyTenantLabel(r.Tenant, labels)
+			if err != nil {
+				return 0, fmt.Errorf("db-ingestor: %w", err)
+			}
+		}
```

Comment:
Important organizational question: with the new refactor, can't this code go into a parser preProcessor? The preprocessor would just modify ts.Labels with the updated labels and return an error an auth error if needed.
That way the ingestor does not need to know anything about auth or multitenancy, nor would the modification to the ingestor's input (The Request change) be necessary.

Reply:
I agree thats gonna be more clean but I fear running an extra loop on incoming timeseries will be costly, than handling all this in Ingest(). Is it right or am I wrong?